politeiawww/invoices_test.go

```diff
@@ -0,0 +1,270 @@
+	invoiceInvalidIndexMimeType := createNewInvoice(t, id,
+		[]www.File{*indexJpeg}, ii.Month, ii.Year)
+
+	// Index file too large test.
+	// It creates a valid line item input, but too large to be accepted
+	lineItemLabor := cms.LineItemsInput{
+		Type:          cms.LineItemTypeLabor,
+		Domain:        "Development",
+		Subdomain:     "politeia",
+		Description:   "PR #2",
+		ProposalToken: "",
+		Labor:         20,
+		Expenses:      0,
+	}
+	tooManyLineItems := make([]cms.LineItemsInput, 0, 5000)
+	for i := 0; i < 5000; i++ {
+		tooManyLineItems = append(tooManyLineItems, lineItemLabor)
+	}
+	invalidInvoiceInput := createInvoiceInput(t, tooManyLineItems)
+	jsonLarge := createInvoiceJSON(t, invalidInvoiceInput)
+	invoiceIndexLarge := createNewInvoice(t, id, []www.File{*jsonLarge},
+		ii.Month, ii.Year)
+
+	// Too many index files test
+	invoiceMaxIndexFiles := createNewInvoice(t, id, []www.File{*json, *json},
+		ii.Month, ii.Year)
+
+	// Attachment file too large test
+	fileLarge := createFilePNG(t, true)
+	invoiceAttachmentLarge := createNewInvoice(t, id,
+		[]www.File{*json, *fileLarge}, ii.Month, ii.Year)
+
+	// Too many attached files test
+	files := make([]www.File, 0, cms.PolicyMaxAttachments+1)
+	files = append(files, *json)
+	for i := 0; i < cms.PolicyMaxAttachments+1; i++ {
+		m := md
+		m.Name = fmt.Sprintf("%v.md", i)
+		files = append(files, m)
+	}
+	invoiceMaxAttachments := createNewInvoice(t, id, files, ii.Month, ii.Year)
+
+	// Setup test cases
+	// XXX this only adds partial test coverage to validateInvoice
+	var tests = []struct {
+		name       string
+		newInvoice cms.NewInvoice
+		user       *user.User
+		want       error
+	}{
+		{"correct invoice", *ni, usr, nil},
+
+		{"invalid signature", invoiceInvalidSig, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusInvalidSignature,
+			}},
+
+		{"incorrect signature", *invoiceIncorrectSig, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusInvalidSignature,
+			}},
+
+		{"no index file", invoiceNoIndexFile, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusNoIndexFile,
+			}},
+
+		{"invalid index mime type", *invoiceInvalidIndexMimeType, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusInvalidIndexFileMimeType,
+			}},
+
+		{"index file too large", *invoiceIndexLarge, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusMaxIndexFileSizeExceeded,
+			}},
+
+		{"too many index files", *invoiceMaxIndexFiles, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusMaxIndexFileExceeded,
+			}},
+
+		{"attachment file too large", *invoiceAttachmentLarge, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusMaxAttachmentSizeExceeded,
+			}},
+
+		{"too many attached files", *invoiceMaxAttachments, usr,
+			www.UserError{
+				ErrorCode: www.ErrorStatusMaxAttachmentsExceeded,
+			}},
+	}
+
+	// Run test cases
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := p.validateInvoice(test.newInvoice, test.user)
+			got := errToStr(err)
```

Comment:
The cms human readable error messages need to be added to the `errToStr` function.

Reply:
Correct way to do this is to pass down pi mode to `errToStr` and switching the api accordingly, since they already exist and should be separate. These tests only use errors related to piwww's api though, so all human readable error messages are added already. this can wait until its needed, or not, i can patch that up as well for this pr.

Reply:
> Correct way to do this is to pass down pi mode to errToStr and switching the api accordingly, since they already exist and should be separate.

huh?

>  These tests only use errors related to piwww's api though

Ok, that's fine then.

Reply:
If we are going to use `errToStr` for both piwww and cmswww tests, it needs to know where to search for those error messages.

Ok cool

Reply:
```
func errToStr(e error) string {
	if e == nil {
		return "nil"
	}

	userErr, ok := e.(www.UserError)
	if ok {
		err, ok := www.ErrorStatus[userErr.ErrorCode]
		if ok {
			return err
		}
		err, ok = cms.ErrorStatus[userErr.ErrorCode]
		if ok {
			return err
		}
	}

	return e.Error()
}

```